WordPress/WordPressNotificationServiceExtension/Sources/NotificationService.swift

```diff
@@ -40,39 +45,37 @@ class NotificationService: UNNotificationServiceExtension {
         let service = NotificationSyncServiceRemote(wordPressComRestApi: api)
         self.notificationService = service
 
-        let identifiers = [ String(noteID) ]
-        service.loadNotes(noteIds: identifiers) { [tracks] error, notifications in
-            defer {
-                contentHandler(notificationContent)
-            }
-
+        service.loadNotes(noteIds: [noteID]) { [tracks] error, notifications in
             if let error = error {
                 tracks.trackNotificationRetrievalFailed(notificationIdentifier: noteID, errorDescription: error.localizedDescription)
                 return
             }
 
-            if let remoteNotifications = notifications,
+            guard
+                let remoteNotifications = notifications,
                 remoteNotifications.count == 1,
                 let notification = remoteNotifications.first,
-                notification.kind == .comment,
-                let body = notification.body,
-                let bodyBlocks = body as? [[String: AnyObject]] {
+                notification.kind == .comment
+            else
+            {
+                return
+            }
 
-                let parser = RemoteNotificationActionParser()
-                let blocks = NotificationContentFactory.content(
-                    from: bodyBlocks,
-                    actionsParser: parser,
-                    parent: notification)
+            let contentFormatter = RichNotificationContentFormatter(notification: notification)
 
-                if let comment: FormattableCommentContent = FormattableContentGroup.blockOfKind(.comment, from: blocks),
-                    let notificationText = comment.text,
-                    !notificationText.isEmpty {
+            guard let bodyText = contentFormatter.formatBody() else { return }
+            notificationContent.body = bodyText
 
-                    notificationContent.body = notificationText
-                }
+            let viewModel = RichNotificationViewModel(
+                attributedBody: contentFormatter.formatAttributedBody(),
```

Comment:
Just curious – is there a reason that the service extension does the formatting of the content for display instead of the content extension?

Reply:
Adapting the `FormattableContent` abstractions for the service extension was a bit challenging. By styling the strings here, we could preserve the architectural boundary a bit.

Reply:
Sounds good to me!